staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go

```diff
@@ -0,0 +1,216 @@
+		names := append([]string{}, certSpec.ips...)
+		names = append(names, certSpec.names...)
+		certProvider, err := createTestTLSCerts(certSpec, names)
+		if err != nil {
+			t.Fatal(err)
+		}
+		sniCerts = append(sniCerts, certProvider)
+	}
+
+	dynamicCertificateController := NewDynamicServingCertificateController(
+		*tlsConfig,
+		&nullCAContent{name: "client-ca"},
+		defaultCertProvider,
+		sniCerts,
+		nil, // TODO see how to plumb an event recorder down in here. For now this results in simply klog messages.
+	)
+	if err := dynamicCertificateController.RunOnce(); err != nil {
+		t.Fatal(err)
+	}
+	tlsConfig.GetConfigForClient = dynamicCertificateController.GetConfigForClient
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	server := &http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		}),
+		MaxHeaderBytes: 1 << 20,
+		TLSConfig:      tlsConfig,
+	}
+	listener, _, err := createListener("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	apiPort := listener.Addr().String()
+	go func() {
+		t.Logf("listening on %s", listener.Addr().String())
+		listener = tls.NewListener(listener, server.TLSConfig)
+		if err := server.Serve(listener); err != nil {
+			t.Error(err)
+		}
+
+		<-stopCh
+		server.Close()
+		t.Logf("stopped listening on %s", listener.Addr().String())
+	}()
+
+	time.Sleep(1 * time.Second)
+
+	expectedDefaultCertBytes, _ := defaultCertProvider.CurrentCertKeyContent()
+	expectedServiceCertBytes, _ := sniCerts[0].CurrentCertKeyContent()
+	expectedLocalhostCertBytes, _ := sniCerts[1].CurrentCertKeyContent()
+	expectedServiceV6CertBytes, _ := sniCerts[2].CurrentCertKeyContent()
+
+	tests := []struct {
+		name       string
+		serverName string
+		expected   []byte
+	}{
+		{
+			name:       "default", // the no server name hits 127.0.0.1, which we hope matches by IP
+			serverName: "",
+			expected:   []byte(strings.TrimSpace(string(expectedLocalhostCertBytes))),
+		},
+		{
+			name:       "invalid",
+			serverName: "not-marked",
+			expected:   []byte(strings.TrimSpace(string(expectedDefaultCertBytes))),
+		},
+		{
+			name:       "service by dns",
+			serverName: "openshift",
+			expected:   []byte(strings.TrimSpace(string(expectedServiceCertBytes))),
+		},
+		{
+			name:       "service v6 by dns",
+			serverName: "openshiftv6",
+			expected:   []byte(strings.TrimSpace(string(expectedServiceV6CertBytes))),
+		},
+		{
+			name:       "localhost by dns",
+			serverName: "localhost",
+			expected:   []byte(strings.TrimSpace(string(expectedLocalhostCertBytes))),
+		},
+		{
+			name:       "localhost by IP",
+			serverName: "127.0.0.1", // this can never actually happen, but let's see
+			expected:   []byte(strings.TrimSpace(string(expectedLocalhostCertBytes))),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, actualDefaultCertBytes, err := cert.GetServingCertificates(apiPort, test.serverName)
```

Comment:
just saw the `func GetServingCertificates` impl... looks like that will make the server log `http: TLS handshake error from 127.0.0.1:62829: EOF` on every invocation... mind spawning an issue/follow-up to improve that?

Reply:
nevermind, I misread the utility method and accidentally caused the issue locally by switching the server to `.ServeTLS` and didn't notice you already had a tls listener